onnxruntime/core/graph/graph_utils.cc

```diff
@@ -77,25 +77,128 @@ bool IsSingleInSingleOutNode(const Node& node) {
+      const auto& subgraph_node_implicit_inputs = subgraph_node.ImplicitInputDefs();
+      if (!subgraph_node_implicit_inputs.empty()) {
+        auto subgraph_node_also_consumes_nodearg_as_implicit_input =
+            std::find_if(subgraph_node_implicit_inputs.cbegin(),
+                         subgraph_node_implicit_inputs.cend(),
+                         [&removed_output_name](const NodeArg* input) {
+                           return input->Name() == removed_output_name;
+                         });
+
+        if (subgraph_node_also_consumes_nodearg_as_implicit_input != subgraph_node_implicit_inputs.cend()) {
+          UpdateImplicitInputNameInSubgraph(subgraph_node, removed_output_name, new_output_name);
+        }
+      }
+
+      // Need mutable input defs to be able to update the implicit input names
+      auto& input_args = subgraph_node.MutableInputDefs();
+
+      if (!input_args.empty()) {
+        int input_slot_index = -1;
+        for (const auto* input_arg : input_args) {
+          ++input_slot_index;
+          // if the input matches, replace the NodeArg with one using the new name
+          if (input_arg->Exists() && input_arg->Name() == removed_output_name) {
+            // sanity check there was no edge for this input. implicit inputs from outer scope do not have edges
+            ORT_ENFORCE(std::count_if(subgraph_node.InputEdgesBegin(), subgraph_node.InputEdgesEnd(),
+                                      [input_slot_index](const Node::EdgeEnd& entry) {
+                                        return entry.GetDstArgIndex() == input_slot_index;
+                                      }) == 0);
+
+            // Create a new NodeArg with the new name
+            input_args[input_slot_index] = &attr_subgraph_pair.second->GetOrCreateNodeArg(new_output_name,
+                                                                                          input_arg->TypeAsProto());
+          }
+        }
+      }
+    }
+  }
+}
+
 bool RemoveSingleInSingleOutNode(Graph& graph, Node& node) {
   if (!IsSingleInSingleOutNode(node)) {
     return false;
   }
+
   // Get input/output edges, nodes, and node args.
   const Node::EdgeEnd& input_edge = *node.InputEdgesBegin();
-  const NodeIndex input_edge_node = input_edge.GetNode().Index();
+  const Node& input_edge_node = input_edge.GetNode();
   const int input_edge_dst_arg = input_edge.GetSrcArgIndex();
   const Node::EdgeEnd& output_edge = *node.OutputEdgesBegin();
-  const NodeIndex output_edge_node = output_edge.GetNode().Index();
+  const Node& output_edge_node = output_edge.GetNode();
   const int output_edge_dst_arg = output_edge.GetDstArgIndex();
 
+  // check if the output from the node to remove is implicit input to the downstream node.
+  // if so, it's used in subgraph in that node and we have to check if it's valid to update that subgraph
+  bool is_implicit_input_to_subgraph = output_edge_dst_arg >= output_edge.GetNode().InputDefs().size();
```

Comment:
I believe so. We have the explicit input defs followed by the implicit ones, so if the index is higher than the number of explicit input defs, it's an implicit input. A helper method in EdgeEnd to check if an input is implicit may be clearer.

Reply:
Sounds good. Yes, I was confused as to whether InputDefs() returns the implicit inputs as well (but I see that it does not).

Reply:
I don't think there's a great place to put this currently. The EdgeEnd itself is generic and whether it's used for an input or output edge is implicit. It would maybe be better to actually tag the EdgeEnd with its type so it could provide this sort of info.

For now I'll add a helper in graph_utils.cc with a bit more explanation. It's a larger piece of restructuring to make this logic about synthetically combining the explicit and implicit inputs more shareable (currently it's hardcoded in places like AddEdge and calls to that).

Reply:
Yeah my hunch was also that EdgeEnd would not have enough info to do that. A helper in graph_utils is good enough.